### ISSUE TITLE
Add "Disk cache size" setting to the Preferences in MacOS

### ISSUE THREAD

**Issue by Jack20947** [transmission/transmission#6673]
### What is the issue?

The default disk cache size (4MB) is small compared to the quantity of RAM and connection speed that we have these days.
My proposal is to increase it to at least 256MB and/or add the possibility to modify it in the Preferences of the app.

This reduce the number of accesses to the disk by Trasnmission and help a lot if the down/upload disk is a mechanical drive

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.5

**Comment by nevack** [comment]
Isn't it settable in `"~/Library/Application Support/Transmission/settings.json"`?
This is an advanced setting, so there's no option to set it with GUI right now.

**Comment by Jack20947** [comment]
I have no settings.json in "~/Library/Application Support/Transmission/".

**Comment by killemov** [comment]
The field is session is `cache-size-mb`. It does not need to be large. I don't know what the ideal size should be. I run it a 16MB, I know some that run it at 0 and I know some that run it at 1GB. I think if the cache is large enough for the piece size, that should be enough. Flushing a large chunk to disk also has some impact.

**Comment by nevack** [comment]
> Do I have to add a in there?

No, you have to create and/or edit `"$HOME/Library/Application Support/Transmission/settings.json"`

Then follow https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md for examples and options.

> And in any case, as I said, it should the exposed

I do not agree.
Not all settings are exposed via GUI.
This is true for GTK and Qt clients as well.

> at least increase the default.

This is debatable, what value must be default.
There's a lot of seeding boxes much less powerful than your machine, Raspberry Pis people run Transmission on wireless routers, etc.

There's no differentiation between libtransmission inside GUI clients and headless daemons.

**Comment by nevack** [comment]
@Jack20947 
I have had some time thinking about how all this can be improved.
Would it be easier for you if there was a button control in app setting, opening this config file location in Finder, and a link (maybe also a button) control, opening https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md doc?

**Comment by Jack20947** [comment]
@nevack

**Comment by nevack** [comment]
> But in case of MacOS it seems to me that the "settings.json" does nothing ... and there is no way to modify _cache-size-mb_ beacuse it's not in the plist (see at the end of the [Editing-Configuration-Files.md](https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md?rgh-link-date=2024-03-07T17%3A18%3A38Z)

@ckerr Charles, I also think this is broken.

The [overload order](https://github.com/transmission/transmission/blob/main/libtransmission/session.cc#L558-L561) is not working correctly.

[Defaults](https://github.com/transmission/transmission/blob/main/macosx/Controller.mm#L459) are overloading any value specified in `settings.json`.

I don't think anyone would like it to work like that.

**Comment by nevack** [comment]
@Jack20947 Thank you for your observation. I'm seeing the same thing, at least at current `main` branch.
This needs to be fixed.

But let's assume everything is working as expected and get back to

> But in case of MacOS it seems to me that the "settings.json" does nothing ... and there is no way to modify cache-size-mb beacuse it's not in the plist (see at the end of the [Editing-Configuration-Files.md](https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md?rgh-link-date=2024-03-07T20%3A20%3A05Z)

`"$HOME/Library/Preferences/org.m0k.transmission.plist"` is for storing macOS client settings.
Like Window states and sizes, GUI state (for example, whether filter bar is shown), location for storing files, etc.
You can inspect all this with `defaults read org.m0k.transmission`

Settings for `libtransmission`, torrent client code itself, are mainly stored in `settings.json`.
If there's no such file in `"$HOME/Library/Application Support/Transmission/"`, this means you are using all compiled in default settings, like `4` for `cache_size_mb`.
You can create this file and adjust anything you see in specified doc page, and Transmission's `libtransmission` will pick it up.

I didn't check this with 4.0.5, it may be not so broken here, maybe give it a try.

**Comment by nevack** [comment]
@tearfur Sorry for bringing you here, but you are pretty active here to not ask you.
Have you observed this behavior on other platforms?

**Comment by Jack20947** [comment]
@nevack 
> I didn't check this with 4.0.5, it may be not so broken here, maybe give it a try.

I'm using 4.0.5 and settings.json doesn't seem to work.

**Comment by tearfur** [comment]
@nevack No, other platforms seems to be working fine.

I have never worked on the macOS code, nor do I have any way to, but I did try to read some of the macOS code before.

It seems to me the macOS client for some reason, since its beginnings, have been duplicating `settings.json` settings inside its plist file, including settings for libtransmission.

It also does not call `tr_sessionSaveSettings()` anywhere in its code. So over the years, any updates we've made to `settings.json`'s format essentially leaves a hole in the macOS app's preferences.

**Comment by tearfur** [comment]
From the limited knowledge I have for the macOS app, it seems to me a better way to handle the preferences would be to use `settings.json` for core settings, and the plist for macOS app specific settings. The main advantage of this is that it saves a lot of headache from trying to keep `settings.json` and the plist file in sync.

**Comment by Coeur** [comment]
In my opinion, in the absence of custom settings, we should call `tr_sys_path_get_capacity` once from libtransmission/file.h, and apply some formula, like:
2 TB disk or more -> 256 MB cache
1 TB disk -> 128 MB cache
500 GB disk -> 64 MB cache
200 GB disk -> 32 MB cache
100 GB disk -> 16 MB cache
50 GB disk -> 8 MB cache
less -> 4 MB cache

**Comment by killemov** [comment]
@Coeur Why? I get that it's at least something, but why use disk size as a metric? Number of downloading torrents and/or largest active piece size would make more sense to me.

**Comment by Jack20947** [comment]
Short update/note: In Transmission 4.0.5 MacOS, it seems that if the Disk cache size is modified (using for example "Transmission Remote GUI" at runtime, the change it taken into account even if it's not retained between restarts.
For example I set the size from 4 to 4096 (I've got a lot of RAM), and both the HD responsivity and the download speed increased dramatically.
So as a workaround, for the moment, I'll do this, but it would be nice to have this fixed eventually.

**Comment by tearfur** [comment]
Putting the whole `Add "Disk cache size" setting to the Preferences` aside, at least we found out we need to add back a bunch of missing settings for macOS app.

The way I see it, the problem here is that a subset of the libtransmission settings is duplicated to the plist file and persistent across sessions, while the rest are not, because the macOS app does not use `tr_sessionSaveSettings()`, so they are not being saved.

The most straightforward way is just to remove the duped libtransmission settings from the plist, and use `settings.json` instead.

**Comment by ckerr** [comment]
> The way I see it, the problem here is that a subset of the libtransmission settings is duplicated to the plist file and persistent across sessions, while the rest are not, because the macOS app does not use tr_sessionSaveSettings(), so they are not being saved.

Yes. The fragmentation between the GTK, Qt, and macOS apps is the project's most serious flaw design flaw. It makes it difficult / impossible to share features between apps and slows down development velocity considerably.

It would be a lot of work, but it would be good to refactor the common pieces into libtransmission-app instead having three separate implementations in the qt/, gtk/, and macos/ directories.

Starting small: app settings seems like a good next step to work on post-4.1.0 since it could build on top of some of last month's serializer code.

**Comment by ckerr** [comment]
I'm going to decline this ticket since we're probably going to just remove the disk cache altogether.

**Comment by killemov** [comment]
> I'm going to decline this ticket since we're probably going to just remove the disk cache altogether.

Has that been decided or still under discussion in #7530?
